src/librustc_typeck/astconv.rs

```diff
@@ -1401,11 +1401,23 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
 
                 assert_eq!(opt_self_ty, None);
                 tcx.prohibit_type_params(&path.segments);
-                let ty = tcx.item_type(def_id);
-                if let Some(free_substs) = self.get_free_substs() {
-                    ty.subst(tcx, free_substs)
+
+                // FIXME: Self type is not always computed when we are here because type parameter
+                // bounds may affect Self type and have to be converted before it.
```

Comment:
I was working on the on-demand stuff and I just realized this can be made to work with `AstConv::get_item_type`.
In all fairness, `ItemImpl` is not hooked up to that in `rustc_typeck::collect` because of predicate complications, that are fixed on this branch I'm playing with (even if you can't get it from `tcx.item_type`, that's not needed here).